jenkins_node_scanner.py

```diff
@@ -0,0 +1,157 @@
+#!/bin/env python3
+
+"""Continually scan a Jenkins master and write the nodes to a file."""
+
+from contextlib import contextmanager
+from xml.etree import ElementTree
+import argparse
+import json
+import logging
+import re
+import sys
+import time
+
+import jenkins
+from prometheus_client import start_http_server, Counter, Gauge, Summary
+
+
+JENKINS_API_LATENCY = Summary(
+    'jenkins_api_latency_seconds',
+    'Response latency of Jenkins API (seconds)',
+    ['api_method'],
+)
+
+JENKINS_API_EXCEPTIONS = Counter(
+    'jenkins_api_exceptions',
+    'Exceptions encountered accessing the Jenkins API.',
+    ['api_method'],
+)
+
+GLOBAL_EXCEPTIONS = Counter(
+    'node_scraper_exceptions',
+    'Unhandled top-level exceptions.',
+)
+
+NODES_FOUND = Gauge(
+    'nodes_found',
+    'Nodes found in a scan.',
+    ['jenkins_master'],
+)
+
+UNPARSEABLE_NODES_FOUND = Gauge(
+    'unparseable_nodes_found',
+    'Nodes found in a scan whose ips could not be parsed.',
+    ['jenkins_master'],
+)
+
+
+@contextmanager
+def ignore_exceptions_except_exit():
+    """Ignores all exceptions except KeyboardInterrupt and SystemExit."""
+    try:
+        yield
+    except (KeyboardInterrupt, SystemExit):
+        raise
+    except:
+        logging.exception('Unexpected error, continuing anyway.')
```

Comment:
Why do we want to continue anyways? Shouldn't we rather abort in such cases?

Reply:
This is a daemon meant to run in the background, so we don't want a network error from Jenkins stopping the whole thing.

Reply:
Good point, nevermind!